examples/wmt/input_pipeline.py

```diff
@@ -0,0 +1,251 @@
+      tf.data.experimental.bucket_by_sequence_length(
+          example_length,
+          boundaries,
+          batch_sizes,
+          pad_to_bucket_boundary=True,
+          drop_remainder=drop_remainder))
+  return dataset
+
+
+def preprocess_wmt_data(dataset,
+                        training,
+                        n_devices,
+                        shuffle_buffer_size=1024,
+                        max_length=512,
+                        batch_size=256,
+                        bucket_length=32,
+                        buckets=None,
+                        drop_remainder=True,
+                        prefetch_size=AUTOTUNE):
+  """Shuffle and batch the given dataset."""
+  buckets = copy.copy(buckets)
+
+  def length_filter(max_len):
+    def filter_fn(x):
+      source, target = x['inputs'], x['targets']
+      l = tf.maximum(tf.shape(source)[0], tf.shape(target)[0])
+      return tf.less(l, max_len + 1)
+    return filter_fn
+
+  if max_length > 0:
+    dataset = dataset.filter(length_filter(max_length))
+
+  if training:
+    dataset = dataset.shuffle(shuffle_buffer_size)
+    dataset = dataset.repeat()
+
+  dataset = bin_and_batch(
+      dataset,
+      n_devices,
+      batch_size=batch_size,
+      bucket_length=bucket_length,
+      drop_remainder=drop_remainder,
+      buckets=buckets)
+
+  if prefetch_size:
+    dataset = dataset.prefetch(prefetch_size)
+
+  return dataset
+
+
+def get_wmt_datasets(n_devices,
+                     dataset_name='wmt17_translate/de-en',
+                     data_dir=None,
+                     vocab_path=None,
+                     target_vocab_size=2**15,
+                     max_corpus_chars=10**7,
+                     batch_size=256,
+                     bucket_length=32,
+                     dynamic_batching=True,
+                     max_target_length=512,
+                     max_eval_target_length=512):
+  """Load and return dataset of batched examples for use during training."""
+  if batch_size % n_devices:
+    raise ValueError("Batch size %d isn't divided evenly by n_devices %d" %
+                     (batch_size, n_devices))
+  if vocab_path is None:
+    vocab_path = os.path.expanduser('~/vocab.subwords')
+
+  train_data = tfds.load(
+      name=dataset_name,
+      split=tfds.Split.TRAIN,
+      data_dir=data_dir,
+      shuffle_files=True)
+
+  eval_data = tfds.load(
+      name=dataset_name,
+      split=tfds.Split.VALIDATION,
+      data_dir=data_dir,
+      shuffle_files=False)
+  features_info = tfds.builder(dataset_name, data_dir=data_dir).info
+
+  # standardize on 'inputs' and 'targets' features.
+  # e.g. {'de': example_in, 'en': example_out} to
+  #      {'inputs': example_in, 'targets': example_out}
+  to_features_dict = (
+      lambda x: {'inputs': x[features_info.supervised_keys[0]],
+                 'targets': x[features_info.supervised_keys[1]]})
+  train_data = train_data.map(to_features_dict)
+  eval_data = eval_data.map(to_features_dict)
```

Comment:
Can add num_parallel_calls here (~2x speedup)

Reply:
yeah this input pipeline is terrible - I'll have a new one coming soon in another PR.